File: LeetCode_Python/binaryTree/889.construct-binary-tree-from-preorder-and-postorder-traversal.py
#
# @lc app=leetcode id=889 lang=python3
#
# [889] Construct Binary Tree from Preorder and Postorder Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def constructFromPrePost(self, preorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        if len(preorder)==0:
            return 
        if len(preorder)==1:
            root = TreeNode(val=preorder[0], left=None, right=None)
            return root
        if len(preorder) != len(postorder):
            raise ValueError('len should be identical')

        rootVal = preorder[0]
        lTreeVal = preorder[1]
        
        for idx, val in enumerate(postorder):
            if val == lTreeVal:
                lTreePostIdx = idx 
                
        # construct postorder_L
        postorder_L = postorder[0:lTreePostIdx+1]
        # construct preorder_L
        preorder_L = preorder[1:1+len(postorder_L)]
        # construct postorder_R
        postorder_R = postorder[lTreePostIdx+1:-1]
        # construct preorder_R
        preorder_R = preorder[1+len(postorder_L):]
        

        # construct preorder_L

        leftRoot = self.constructFromPrePost(preorder_L, postorder_L)
        rightRoot = self.constructFromPrePost(preorder_R, postorder_R)

        root = TreeNode(val=rootVal, left=leftRoot, right=rightRoot)
        return root

        
        
# @lc code=end

File: LeetCode_Go/BinaryTree/889.construct-binary-tree-from-preorder-and-postorder-traversal.go
/*
 * @lc app=leetcode id=925 lang=golang
 *
 * [889] Construct Binary Tree from Preorder and Postorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// build a binary tree from postorder and inorder traversal integer arrays. There are multiple solutions, return any of them.

// Given a tree in preorder, the first number is the root. The second number is the left subtree's root, and also the last number in left subtree's slice in the postorder. (1|245|36, 452|63|1, look for `2`). use the left subtree root to find the boundary of left subtree and right subtree in the two arrays. and recursively build the left and right subtrees. Use a hashmap to store inorder index for faster slicing inorder.

// O(n) time | O(n) space

func constructFromPrePost(preorder []int, postorder []int) *TreeNode {
	postorderIdxMap := make(map[int]int)
	for idx, val := range postorder {
		postorderIdxMap[val] = idx
	}
	return buildSubTree(preorder, postorder, 0, len(preorder)-1, 0, len(postorder)-1, postorderIdxMap)
}

func buildSubTree(preorder, postorder []int, preLeft, preRight, postLeft, postRight int, postorderIdxMap map[int]int) *TreeNode {
	if postLeft > postRight || preLeft > preRight {
		return nil
	}

	// build root
	root := &TreeNode{Val: preorder[preLeft]}
	if preLeft == preRight { // single node case
		root.Left, root.Right = nil, nil
		return root
	}
	// find slicing boundaries
	leftRootVal := preorder[preLeft+1] // assume the second number is the left subtree's root. But left subtree's root can also be nil, which make the solution not unique.

	// postorder
	leftTreePostLeft := postLeft
	leftTreePostRight := postorderIdxMap[leftRootVal]
	rightTreePostLeft := postorderIdxMap[leftRootVal] + 1
	rightTreePostRight := postRight - 1
	// preorder
	leftTreePreLeft := preLeft + 1
	leftTreePreRight := preLeft + 1 + leftTreePostRight - leftTreePostLeft
	rightTreePreLeft := preRight - (rightTreePostRight - rightTreePostLeft)
	rightTreePreRight := preRight

	// build subtrees
	// connect to root
	root.Left = buildSubTree(preorder, postorder, leftTreePreLeft, leftTreePreRight, leftTreePostLeft, leftTreePostRight, postorderIdxMap)
	root.Right = buildSubTree(preorder, postorder, rightTreePreLeft, rightTreePreRight, rightTreePostLeft, rightTreePostRight, postorderIdxMap)

	return root
}

// @lc code=end

